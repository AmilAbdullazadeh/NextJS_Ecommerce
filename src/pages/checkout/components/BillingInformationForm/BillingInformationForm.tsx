import React, { useState, useEffect } from 'react';
import { TextField } from '@material-ui/core';
import Grid from '@material-ui/core/Grid';

import {
    useDebounce,
    useAppDispatch,
    useAppSelector as useSelect,
} from '@hooks';
import {
    billingAddressSelector,
    billingFullNameSelector,
    isValidBillingAddressSelector,
    isValidBillingFullNameSelector,
} from '@pages/checkout/store/selectors';
import {
    updateBillingAddressAction,
    updateBillingFullNameAction,
} from '@pages/checkout/store/actions';

const BillingInformationForm: React.FC = () => {
    const inputDebounce = 250;
    const dispatch = useAppDispatch();

    const billingFullName = useSelect(billingFullNameSelector);
    const billingAddress = useSelect(billingAddressSelector);

    const isValidBillingFullName = useSelect(isValidBillingFullNameSelector);
    const isValidBillingAddress = useSelect(isValidBillingAddressSelector);

    const [billingInfo, setBillingInfo] = useState({
        fullName: billingFullName,
        address: billingAddress,
    });

    const [fullNameError, setFullNameError] = useState(false);
    const [addressError, setAddressError] = useState(false);

    const debouncedBillingFullName = useDebounce(
        billingInfo.fullName,
        inputDebounce
    );

    const debouncedBillingAddress = useDebounce(
        billingInfo.address,
        inputDebounce
    );

    useEffect(() => {
        dispatch(
            updateBillingFullNameAction({ fullName: debouncedBillingFullName })
        );
    }, [debouncedBillingFullName]);

    useEffect(() => {
        dispatch(
            updateBillingAddressAction({ address: debouncedBillingAddress })
        );
    }, [debouncedBillingAddress]);

    const handleFullNameChange = (e: React.ChangeEvent<HTMLInputElement>) =>
        setBillingInfo(prevInfo => ({
            ...prevInfo,
            fullName: e.target.value,
        }));

    const handleAddressChange = (e: React.ChangeEvent<HTMLInputElement>) =>
        setBillingInfo(prevInfo => ({
            ...prevInfo,
            address: e.target.value,
        }));

    const textFieldProps = [
        {
            label: 'full name',
            onChange: handleFullNameChange,
            value: billingInfo.fullName,
            onBlur: () => setFullNameError(!isValidBillingFullName),
            error: fullNameError,
            helperText: fullNameError
                ? 'field should contain a valid full name'
                : '',
        },
        {
            label: 'address',
            onChange: handleAddressChange,
            value: billingInfo.address,
            onBlur: () => setAddressError(!isValidBillingAddress),
            error: addressError,
            helperText: addressError
                ? 'field should contain a valid address'
                : '',
        },
    ];

    return (
        <div id="billing-info-form">
            <Grid
                container
                direction="column"
                justifyContent="center"
                alignItems="center"
                spacing={2}
            >
                {textFieldProps.map(({ label, onChange, value, ...rest }) => (
                    <Grid item xs={12} key={label}>
                        <TextField
                            label={label}
                            variant="outlined"
                            color="primary"
                            onChange={onChange}
                            inputProps={{
                                'aria-label': label,
                                'data-testid': label,
                            }}
                            value={value}
                            {...rest}
                        />
                    </Grid>
                ))}
            </Grid>
        </div>
    );
};

export default BillingInformationForm;
