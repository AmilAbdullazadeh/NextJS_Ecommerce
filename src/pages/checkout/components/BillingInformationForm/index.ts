export { default as BillingInformationForm } from './BillingInformationForm';
