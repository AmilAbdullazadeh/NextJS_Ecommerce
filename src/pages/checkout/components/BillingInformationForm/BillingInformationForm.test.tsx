import React from 'react';
import faker from 'faker';

import { renderWithProviders, fireEvent, waitFor } from '@testUtils';

import BillingInformationForm from './BillingInformationForm';

describe('BillingInformationForm', () => {
    it('renders correctly', () => {
        const {
            container: { firstChild },
        } = renderWithProviders(<BillingInformationForm />);

        expect(firstChild).toMatchSnapshot();
    });

    it('updates billing full name', () => {
        const { getByRole } = renderWithProviders(<BillingInformationForm />);
        const fullNameInput = getByRole('textbox', {
            name: /full name/i,
        });

        const rndFullName = `${faker.name.firstName()} ${faker.name.lastName()}`;

        fireEvent.change(fullNameInput, { target: { value: rndFullName } });

        expect(fullNameInput).toHaveAttribute('value', rndFullName);
    });
    it('shows error text for invalid full name input', async () => {
        const { getByRole, getByText } = renderWithProviders(
            <BillingInformationForm />
        );
        const fullNameInput = getByRole('textbox', {
            name: /full name/i,
        });

        fullNameInput.focus();

        fullNameInput.blur();

        await waitFor(() =>
            expect(
                getByText('field should contain a valid full name')
            ).toBeVisible()
        );
    });

    it('updates billing address', () => {
        const { getByRole } = renderWithProviders(<BillingInformationForm />);
        const addressInput = getByRole('textbox', { name: /address/i });

        const rndAddress = faker.address.streetName();

        fireEvent.change(addressInput, { target: { value: rndAddress } });

        expect(addressInput).toHaveAttribute('value', rndAddress);
    });
    it('shows error text for invalid full name input', async () => {
        const { getByRole, getByText } = renderWithProviders(
            <BillingInformationForm />
        );
        const addressInput = getByRole('textbox', { name: /address/i });

        addressInput.focus();

        addressInput.blur();

        await waitFor(() =>
            expect(
                getByText('field should contain a valid address')
            ).toBeVisible()
        );
    });
});
