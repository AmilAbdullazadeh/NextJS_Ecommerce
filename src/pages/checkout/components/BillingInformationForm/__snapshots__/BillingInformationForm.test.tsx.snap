// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillingInformationForm renders correctly 1`] = `
<div
  id="billing-info-form"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          full name
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            aria-label="full name"
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="full name"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                full name
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
        >
          address
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            aria-label="address"
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="address"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                address
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
  </div>
</div>
`;
